### PR TITLE
Pivot - remove deprecated lifecycle method

### LINKED ISSUE #4613: React.StrictMode compliance
**Maintainer note: see [this GitHub project](https://github.com/microsoft/fluentui/projects/35) for progress on strict mode work, which will be released with the [`@fluentui/react` version 8 release](https://github.com/microsoft/fluentui/wiki/Version-8:-Release-notes).**

---

### Bug Report

- __Package version(s)__:
  - office-ui-fabric-react 5.83.0
  - react 16.3.1

#### Priorities and help requested (not applicable if asking question):

Are you willing to submit a PR to fix? No (Due to lack of time and temporary CLA difficulties)

Requested priority: High 

#### Describe the issue:

Mounting Fabric components in a [StrictMode](https://reactjs.org/blog/2018/03/29/react-v-16-3.html#strictmode-component) React tree causes many warnings in the console, due to [unsafe lifecycle methods](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html) which will be deprecated in a future 16.x release and removed completely with React 17. Warnings concerning [string refs](https://reactjs.org/blog/2018/03/29/react-v-16-3.html#createref-api) are also logged.

This is caused by e.g. the Fabric component, CommandBar, ActionButton, BaseButton, Breadcrumb, DialogBase, FocusZone, ... (lots and lots of components).

I didn't find any issue concerning this, so I thought I'd give a heads up about the issue, since this will have to be fixed eventually if React 17 is to be supported by Fabric.

---

**UPDATE 10/24/19: Some but not all of these issues have been addressed. Please see https://github.com/OfficeDev/office-ui-fabric-react/issues/4613#issuecomment-546165430 and following comments for current status.**

### PR DESCRIPTION
**Pull request by jdhuntington**
#### Pull request checklist

- [X] Addresses an existing issue: Fixes #4613 
- [X] Include a change request file using `$ npm run change`

#### Description of changes

* Remove pivot's use of `componentWillReceiveProps`
* Remove `links` from state
* Render selected pivot item deterministically **(NB: Does change previous unintended behavior.)**

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/8037)

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/Pivot/Pivot.base.tsx [hunk 1]
```diff
@@ -105,23 +81,44 @@ export class PivotBase extends BaseComponent<IPivotProps, IPivotState> {
   }
 
   public render(): JSX.Element {
+    const linkCollection = this._getPivotLinks(this.props);
+    const selectedKey = this._getSelectedKey(linkCollection);
```
Comment (JasonGore) [inline packages/office-ui-fabric-react/src/components/Pivot/Pivot.base.tsx]: This *feels* so much better than having it in state. 👍

File: packages/office-ui-fabric-react/src/components/Pivot/Pivot.base.tsx [hunk 2]
```diff
@@ -34,8 +39,6 @@ const PivotItemType = (<PivotItem /> as React.ReactElement<IPivotItemProps>).typ
  *     </Pivot>
  */
 export class PivotBase extends BaseComponent<IPivotProps, IPivotState> {
-  private _keyToIndexMapping: { [key: string]: number };
-  private _keyToTabIds: { [key: string]: string };
```
Comment (JasonGore) [inline packages/office-ui-fabric-react/src/components/Pivot/Pivot.base.tsx]: 👍

**Comment by jdhuntington** [comment]
Thanks! In profiling it, it didn't look like anything was more expensive than before.
With this change, we do bite off a bit more work in the event handler (recomputing the links), but I think the tradeoff in favor of cleaner state management is worth it.

**Comment by dzearing** [comment]
@ecraig12345 I've discussed with JD, and I believe that if we have `defaultSelectedKey` AND `selectedKey`, there is a clear bug if `selectedKey` does not behave as documented/expected, and if users are treating the bug as a feature, there is a clear path to fix (use `defaultSelectedKey`.)

I would like to merge this.

**Pull request merged**